File: Hackerrank C/Hackerrank C/sherlock_and_watson.c
//
//  sherlock_and_watson.c
//  Hackerrank C
//
//  Created by Dhruv Manchala on 11/8/14.
//  Copyright (c) 2014 Dhruv Manchala. All rights reserved.
//

#include <stdio.h>

void reverse(int* A, int start, int end)
{
    for (int i = 0; i < (end - start + 1) / 2; i++)
    {
        int temp = A[start + i];
        A[start + i] = A[end - i];
        A[end - i] = temp;
    }
}

void rotate(int* A, int len, int n)
{
    n = n % len;
    
    reverse(A, 0, len - 1 - n);
    reverse(A, len - n, len - 1);
    reverse(A, 0, len - 1);
}

int sherlock_and_watson(void)
{
    int n;
    int k;
    int q;
    
    scanf("%d %d %d", &n, &k, &q);
    
    int A[n];
    
    for (int i = 0; i < n; i++) scanf("%d", &A[i]);
    
    rotate(A, n, k);
    
    for (int i = 0; i < q; i++)
    {
        int ind;
        
        scanf("%d", &ind);
        printf("%d\n", A[ind]);
    }
    
    return 0;
}

File: Hackerrank Scala/src/SherlockAndWatson.scala
import scala.runtime.ScalaRunTime.stringOf
import scala.io.StdIn.readLine

object SherlockAndWatson {
  
  def main(args: Array[String]) {

    val Array(_, k, q) = readLine.split(" ").map(_.toInt)
    var A = readLine.split(" ").map(_.toInt)
    
    rotate(A, k)
    
    for (i <- 0 to q - 1) {
    	val ind = readLine.toInt
    	println(A(ind))
    }
  }
  
  def reverse(A: Array[Int], start: Int, end: Int) {
    
    if (start >= end) return
    else {      
      
      val temp = A(start)
      A(start) = A(end)
      A(end) = temp
      
      reverse(A, start + 1, end - 1)
    }
  }
  
  def rotate(A: Array[Int], k: Int) {
    
    val len = A.length
    val k_ = k % len
    
    reverse(A, 0, len - 1 - k_)
    reverse(A, len - k_, len - 1)
    reverse(A, 0, len - 1)
  }
}